test/features/test_organized_edge_detection.cpp

```diff
@@ -0,0 +1,132 @@
+    // organized edge detection algorithm.
+    for (auto row = 0; row < OUTER_SQUARE_EDGE_LENGTH; ++row) {
+      for (auto col = 0; col < OUTER_SQUARE_EDGE_LENGTH; ++col) {
+        float x = col - (OUTER_SQUARE_EDGE_LENGTH / 2);
+        float y = row - (OUTER_SQUARE_EDGE_LENGTH / 2);
+
+        const auto outer_square_ctr = OUTER_SQUARE_EDGE_LENGTH / 2;
+        const auto inner_square_ctr = INNER_SQUARE_EDGE_LENGTH / 2;
+
+        auto depth = SYNTHETIC_CLOUD_BASE_DEPTH;
+        auto r = 0;
+        auto g = 255;
+
+        // If pixels correspond to smaller box, then set depth and color appropriately
+        if (col > outer_square_ctr - inner_square_ctr &&
+            col < outer_square_ctr + inner_square_ctr) {
+          if (row > outer_square_ctr - inner_square_ctr &&
+              row < outer_square_ctr + inner_square_ctr) {
+            depth = SYNTHETIC_CLOUD_BASE_DEPTH - SYNTHETIC_CLOUD_DEPTH_DISCONTINUITY;
+            r = 255;
+            g = 0;
+          }
+        }
+
+        organized_test_cloud->at(col, row).x = x * SYNTHETIC_CLOUD_RESOLUTION;
+        organized_test_cloud->at(col, row).y = y * SYNTHETIC_CLOUD_RESOLUTION;
+        organized_test_cloud->at(col, row).z = depth;
+        organized_test_cloud->at(col, row).r = r;
+        organized_test_cloud->at(col, row).g = g;
+        organized_test_cloud->at(col, row).b = 0;
+        organized_test_cloud->at(col, row).a = 255;
+      }
+    }
+
+    return organized_test_cloud;
+  }
+};
+} // namespace
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+/*
+This test is designed to ensure that the organized edge detection appropriately
+classifies occluding and occluding edges and to specifically detect the type of
+regression detailed in PR 4275 (https://github.com/PointCloudLibrary/pcl/pull/4275).
+This test works by generating a synthetic cloud of one square slightly in front of
+another square, so that occluding edges and occluded edges are generated.  The
+regression introduced in PCL 1.10.1 was a logic bug that caused both occluding and
+occluded edges to be miscategorized as occluding edges.  This test should catch
+this and similar bugs.
+*/
+TEST_F(OrganizedPlaneDetectionTestFixture, OccludedAndOccludingEdges)
+{
+  const auto MAX_SEARCH_NEIGHBORS = 50;
+  const auto DEPTH_DISCONTINUITY_THRESHOLD = SYNTHETIC_CLOUD_DEPTH_DISCONTINUITY / 2.1f;
+
+  // The expected number of occluded edge points is the number of pixels along the
+  // perimeter of the inner square, subject to certain conditions including that the
+  // inner square edge length is an even number.  The expected number of occluding edge
+  // points is similar, but as empirically determined to be 8 less than the number
+  // of perimeter points.
+  const int EXPECTED_OCCLUDING_EDGE_POINTS = INNER_SQUARE_EDGE_LENGTH * 4 - 8;
+  const int EXPECTED_OCCLUDED_EDGE_POINTS = INNER_SQUARE_EDGE_LENGTH * 4;
```

Comment:
Note to self: some of these values still don't make sense to me. Given the discrete nature of the point cloud, I would expect the inner and outer perimeters to be given by
```cpp
// inner perimeter
// disregarding one corner point for each side of the square
(INNER_SQUARE_EDGE_LENGTH - 1) * 4  

// outer perimeter
// the perimeter enclosing the inner square
(INNER_SQUARE_EDGE_LENGTH + 1) * 4;
```

Reply:
Tomorrow I will visualize the edge points and attempt to reconcile them with the logic.  The expected value calculation as written seems to only work with an even number for the inner square length.

Reply:
You are correct.  My cloud generation function was making an inner square that was 49 x 49 rather than 50 x 50.